### PR TITLE
Implemented PAC validation

### PR DESCRIPTION
**Pull request by AleksSavelev**

### CODE REVIEW

**Review by uve — COMMENTED** [review]
1) make functions decompositions
2) move to TypeScipt
3) Generate Custom exceptions
4) use pipe style
5) Improve naming
6) Splint validation logic between multiple small files.

**Review by uve — COMMENTED** [review]
1) make functions decompositions
2) move to TypeScipt
3) Generate Custom exceptions
4) use pipe style
5) Improve naming
6) Splint validation logic between multiple small files.

**Review by uve — COMMENTED** [review]
1) make functions decompositions
2) move to TypeScipt
3) Generate Custom exceptions
4) use pipe style
5) Improve naming
6) Splint validation logic between multiple small files.

**Comment by AleksSavelev** [comment]
Waits for #467 to be merged

File: src/features/FeatureTypes.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    Warning = "warning",
+    Info = "info",
+}
+
+export enum Stage {
+    PreBuild = "pre-build",
+    PostBuild = "post-build",
+}
+
+const DEFAULT = 1;
```
Comment (uve) [inline src/features/FeatureTypes.ts:13]: replace it with enums:
    export const enum VisualUpdateType {
        Default= 1 << 1,
        Slicer = 1 << 2,
        Matrix = 1 << 3,
        All = Default | Slicer | Matrix
    }

File: src/features/FeatureTypes.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+const SLICER = 2;
+const MATRIX = 4;
+
+export enum VisualFeatureType {
+    Default = DEFAULT,
+    Slicer = SLICER | MATRIX,
```
Comment (uve) [inline src/features/FeatureTypes.ts:19]: How does it work: Slicer = SLICER | MATRIX ?

File: src/Package.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        this.sourceCode = sourceCode;
+        this.capabilities = capabilities;
+        this.visualFeatureType = visualFeatureType;
+    }
+    
+    public contain(keyword: string) {
+        return this.sourceCode.indexOf(keyword) > -1;
+    }
+
+    public capabilityEnabled(expectedObject: object) {
```
Comment (uve) [inline src/Package.ts:49]: rename to IsCapabilityEnabled

**Review by uve — CHANGES_REQUESTED** [review]

File: src/features/ColorPalette.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+import BaseFeature from "./BaseFeature.js";
+import { Severity, Stage, VisualFeatureType } from "./FeatureTypes.js";
+
+export default class ColorPalette implements BaseFeature {
+    public static featureName = "Color Palette"
+    public static documentationLink = "https://learn.microsoft.com/en-us/power-bi/developer/visuals/add-colors-power-bi-visual"
+    public static severity = Severity.Warning
+    public static stage = Stage.PostBuild
+    public static visualFeatureType = VisualFeatureType.All
+    public static errorMessage = `${this.featureName} - ${this.documentationLink}`
```
[the remark this answers is not in the document]
Comment (AleksSavelev) [reply src/features/ColorPalette.ts:11]: It's always the same, but it refers to variables that are not declared in BaseFeature. So we can't access them from the base class. It's possible to make it inside the constructor, but I don't think this is a good way.

File: src/features/BaseFeature.ts [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+export default abstract class BaseFeature {
+    public static severity: Severity
+    public static stage: Stage
+    public static visualFeatureType: VisualFeatureType
+    public static featureName: string
+    public static documentationLink: string
+    public static errorMessage: string
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    static isSupported(source: Visual | Package) {}
```
[the remark this answers is not in the document]
Comment (AleksSavelev) [reply src/features/BaseFeature.ts:14]: We have @typescript-eslint/no-unused-vars rule in our eslint, it doesn't allow us to declare the source parameter here. Do we really want to disable it for the whole project?

File: src/features/ColorPalette.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (AleksSavelev) [reply src/features/ColorPalette.ts]: Decided to make errorMessage as getter

**Review by uve — APPROVED** [review]

**Pull request merged**